python/py_measure/__init__.py

```diff
@@ -0,0 +1,118 @@
+
+def result_to_dict(lib, result: ctypes.POINTER(None)) -> "str | dict":
+    """
+    Takes a reference to the library and a pointer to a results object and recursively translates it into a python object.
+    """
+    value = ctypes.c_void_p()
+    if (lib.mapiResultGetValue(result, ctypes.pointer(value))):
+        return ctypes.cast(value, ctypes.c_char_p).value.decode('ascii')
+    else:
+        num = lib.mapiResultGetEntries(result, None, 0)
+        buf = (MapiResultEntry*num)()
+        lib.mapiResultGetEntries(result, buf, num)
+        return {entry.name.decode('ascii'): result_to_dict(lib, entry.value) for entry in buf}
+
+ALLOWED_PROVIDERS = ("git", "system", "energy", "gpu")
+
+class Environment():
+    def __init__(self, providers=ALLOWED_PROVIDERS, verbose=False):
+        self.libmeasureapi = ctypes.cdll.LoadLibrary(path_to_libmeasureapi())
+
+        ## Tell ctypes the datatypes of the respective functons
+        ## NOTE: leaving this out has caused issues with downcasted addresses for the handle in the past.
+        self.libmeasureapi.mapiStartMeasure.argtypes = [MapiConfig]
+        self.libmeasureapi.mapiStartMeasure.restype = ctypes.c_void_p
+        self.libmeasureapi.mapiStopMeasure.argtypes = [ctypes.c_void_p]
+        self.libmeasureapi.mapiStopMeasure.restype = ctypes.c_void_p
+        self.libmeasureapi.mapiSetLogCallback.argtypes = [ctypes.c_void_p]
+        self.libmeasureapi.mapiResultGetValue.argtypes = [ctypes.c_void_p, ctypes.POINTER(ctypes.c_void_p)]
+        self.libmeasureapi.mapiResultGetValue.restype = ctypes.c_bool
+        self.libmeasureapi.mapiResultGetEntries.argtypes = [ctypes.c_void_p, ctypes.c_void_p, ctypes.c_size_t]
+        self.libmeasureapi.mapiResultGetEntries.restype = ctypes.c_size_t
+        self.libmeasureapi.mapiResultFree.argtypes = [ctypes.c_void_p]
+
+        self.handle = None
+        self.providers = self.allowed_providers(providers)
+        self.measurements = []
+        self.verbose = verbose
+
+        if self.verbose:
+            self.libmeasureapi.mapiSetLogCallback(log_callback)
+
+    def allowed_providers(self, providers):
+        ret = []
+        for i in providers:
+            if i not in ALLOWED_PROVIDERS:
+                raise ValueError(f'Unallowed provider: "{i}".')
+            ret.append(i.encode())
+        ret.append(None)
+        return ret
+    
+    def __enter__(self):
+        self.start_measuring()
+        return self
+
+    def __exit__(self, *args):
+        self.stop_measuring()
+
+    def measure(self, func=None):
+        if func:
+            def wrapper(*arg):
+                with self.measure():
+                    return func(*arg)
+            return wrapper
+
+        return self
+
+    def start_measuring(self):
+        if self.handle:
+            raise ValueError('Measuring was alraedy started...')
+
+        providers_native = (ctypes.c_char_p*len(self.providers))(*self.providers)
```

Comment:
The providers must end with a `None` as a "sentinel value"
```suggestion
        providers_native = (ctypes.c_char_p*(len(self.providers)+1))(*self.providers, None)
```

Reply:
Ah, I see, this is already implemented, but in a different location.